nixForLinking: init

9999years opened the pull request:
Overlays for CppNix nightly, Lix, or Tvix want to change the default Nix implementation in Nixpkgs by overriding `pkgs.nix`. However, some packages link against the internal/unstable CppNix APIs directly, and these packages will break if built with different versions or implementations of Nix.

If you want to swap out the Nix implementation in your package set, you don't want these packages to break. Therefore, some packages will refer to `nixForLinking` explicitly, at least until these dependencies can be sorted out.

The addition of an explicit `nixForLinking` attribute decoupled from `nix`, which is just "a Nix implementation", will help Nix implementation maintainers test Nix implementations in Nixpkgs with minimal hassle.



<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

9999years committed:
nixForLinking: init

Overlays for CppNix nightly, Lix, or Tvix want to change the default Nix
implementation in Nixpkgs by overriding `pkgs.nix`. However, some
packages link against the internal/unstable CppNix APIs directly, and
these packages will break if built with different versions or
implementations of Nix.

If you want to swap out the Nix implementation in your package set, you
don't want these packages to break. Therefore, some packages will refer
to `nixForLinking` explicitly, at least until these dependencies can be
sorted out.

The addition of an explicit `nixForLinking` attribute decoupled from
`nix`, which is just "a Nix implementation", will help Nix
implementation maintainers test Nix implementations in Nixpkgs with
minimal hassle.

RaitoBezarius commented:
Let's wait for a Nix maintenance team person to validate this change before the merge. I pinged their team.

roberth reviewed (dismissed):
I'd like to discuss this with the rest of the package maintainers / Nix team.
A possible alternative that's worth considering is to add a `nix-cli` alias which represents exactly that part and not the greater interface area that the Nix _implementation_ has, including the "unstable" C++ API and the in progress C API.

Aleksanaa commented:
You probably know that Nix team is not okay with the name "cppnix". I don't make moral evaluations here, but I want to say that if you want to get things done easier, don't put a greater context into this discussion again. Find an alias that they and we are probably satisfied with.

Ericson2314 commented:
> If you want to swap out the Nix implementation in your package set, you probably don't want to sign up to rebuild Chromium. 

Wait a minute, it doesn't do this. In that chain of derivations, there is at least one fixed-output derivation. Overriding `nix` therefore doesn't cause a rebuild of chromium.

Ericson2314 reviewed (dismissed):
We need some examples of actual non-trivial rebuilds

9999years commented:
> We need some examples of actual non-trivial rebuilds

@Ericson2314 Sure, I can update the comments. `nix-init`, `nurl`, `nixd`, and `nix-du` all rely on the C++ Nix API, so those still need to point at upstream stable Nix.

9999years commented:
@edolstra Thanks for the feedback! I'm happy to rename the attribute if you or another Nix team member would suggest a name for it. Would `nixForLinking` be OK?

9999years commented:
I've renamed the `cppnix` attribute to `nixForLinking`. If that's not sufficient, please suggest an alternate name. I've also removed the references to Chromium in my comments and commit messages.

Ericson2314 commented:
The metapackage/alias also needs a clear contract as to what it implements. E.g. I assume in most cases it is an unspecified portion of the CLI, and thus Tvix would *not* make sense as possible choice of implementation, since (as far as I know) it doesn't aim to implement the same CLI.

llakala commented:
> At a minimum, `nix` should be Nix, and the alias/metapackage should instead get the fresh new name.

I don't believe anyone was suggesting anything but this.

RaitoBezarius commented:
> (as far as I know) it doesn't aim to implement the same CLI.

That's beyond the point, I always planned to implement bits of compatible CLI with Tvix. It would be nice if we could do small progress rather than attempting to solve a full problem space without even a clear idea of what is in scope and what is not.

There's an existing usage of the Nix ABI/API, we would like in this PR to solve the problem. Can we keep more granular ABI/API in further PRs? I don't think we have a strong opinion on whether it should be cut into `nixForLinking`, `nixForNewStyleCli`, `nixForOldStyleCli`, `nixForUnstableCppAPI`, `nixForInProgressCAPI`. `nixForLinking` is a good start IMHO and I would appreciate it if any Nix maintainer could confirm this fresh name is sufficient for them.

Ericson2314 commented:
I'm asking what the meta-package is for, but you have answered the opposite question, which is when when do package need exactly Nix stable.

I understand *that* -- our C++ API is explicitly unstable between major versions, so sure, packages depending on it need to pin specific versions. That's fine. We do this all the time in Nixpkgs.

The meta-package, to recap, is the alias that may be set to either Nix or Lix without breaking things. This is implicit in this PR current because all the remaining usages of `nix` do not show up in the list. I want to know what the contract is for when Nix or Lix should be used. I want a whitelist of when you can use either version, not a blacklist of when you can't.

This --- giving the meta-package the fresh name, given the implementations their own chosen names --- is standard practice in Nixpkgs and other distros. For example, we don't override `gcc` because we might use `clang`, we override `stdenv.cc`. Likewise, I made up the silly name `bintools` almost a decade ago to make clear it is not always GNU Binutils.

Ericson2314 commented:
Small addendum: I just went to double check the debian docs, and it appears that they use "metapackage" for any package which just exists to bring in other packages. Perhaps it is not such a good word for what I mean, where the emphasis on on "merely being an alias" but in particular an alias with a *swapable* definition.

RaitoBezarius commented:
> I'm asking what the meta-package is for, but you have answered the opposite question, which is when when do package need exactly Nix stable.
> 
> I understand _that_ -- our C++ API is explicitly unstable between major versions, so sure, packages depending on it need to pin specific versions. That's fine. We do this all the time in Nixpkgs.
> 
> The meta-package, to recap, is the alias that may be set to either Nix or Lix without breaking things. This is implicit in this PR current because all the remaining usages of `nix` do not show up in the list. I want to know what the contract is for when Nix or Lix should be used. I want a whitelist of when you can use either version, not a blacklist of when you can't.
> 
> This --- giving the meta-package the fresh name, given the implementations their own chosen names --- is standard practice in Nixpkgs and other distros. For example, we don't override `gcc` because we might use `clang`, we override `stdenv.cc`. Likewise, I made up the silly name `bintools` almost a decade ago to make clear it is not always GNU Binutils, but might be Apple cctools (or since then, also LLVM's tools, Mold, etc.).

Not to pile up on @lf- comment which I think summarized it well. Let me rephrase how I understand the current situation.

We are defining the allow-list as a **fuzzy** set of capabilities which are made available via the CLI and possibly some "obvious" C++ API pieces that are fairly stable (even though the C++ API is unstable), we cannot say here's X, Y, Z. We aren ot at this step of the process. We want to get the ball moving, so we are proposing to start by the complementary of the disallow-list.

That complementary may be wrong for now, but we will refine it as we go and as we learn about the consumers of those ABI/APIs.

So to sum it up, we deem it **out of scope** to define this allow-list in this PR. I don't see why this should be a blocker for this PR, the status quo was *NO*thing, we are going to *SOME*thing, it's not *EVERY*thing yet, but if you care about it, we should be able to settle it in further PRs.

Thank you for your understanding.

Ericson2314 commented:
@lf- 

> but it requires that we actually have code review standards to ensure that people don't incorrectly force the nix implementation to one implementation or another when they think "i want a nix"

Yes, but I think that is just life. It's not good to kick Nix off the "Nix" name because we lack static analysis tools.

> and making that also happen for external projects is a tall order.

Again, that's just life I think. If we need better opt-in diagnostic tools around changing based practices regarding metapackages, that is also a general issue worth solving, not something to be solved with a one-off policy exception.

> Regardless of whether we have this as an "allow-list" or a "deny-list" implementation in the end, we need to have the packages that *cannot be expected to build with nix `master`* distinguished from any others, so that nix master can be tested by replacing pkgs.nix.

I don't think this problem has anything to do with Lix vs Nix substitutability. Any such package should depend on `nix_x_y`. This is, again, standard practice for packages whose reverse dependencies require explicit versions.

(The long term solution is either getting these packages off the C++ ABI, or getting Nix developers working on them too, like me on Hydra. Same thing goes for the versions of `nix-eval-jobs` and maybe `hydra` in the Lix monorepo, of course; neither of us claim to have a stable C++ API.)

> this PR is a first step in that future by fixing the problem that you cannot test nix master on client systems that install a pile of tools even if those tools want nixos/nix specifically as pkgs.nix.

I don't think so. I think the first step is pinning Nix versions in projects that require them, per the above.

------

@RaitoBezarius 

> We want to get the ball moving, so we are proposing to start by the complementary of the disallow-list.

My understanding of standard Nixpkgs practices this is a very peculiar way to start the ball rolling. Standard Nixpkgs practices are:

 - Pin specific versions
 - Introduce new names for intensionally-defined meta packages

for getting the ball rolling

In fact, It is perfectly fine if the intensional definition is WIP while it is seen empirically what parts of the interface things are using, in fact. The provisional WIP nature just needs to be the documented, with the idea that as the whitelist grows, the intensional definition is clarified.

Ericson2314 commented:
Plugins should be solved with `nix.withPlugins` or similar. A `nixForLinking` is not good enough because if anyone changes their Nix to be something else (for sake of argument, imagine *NixOS* setting), chaos will still reign. The choice of the libraries that the executable and plugins share must be as "bound as late" as possible.

However, when linking is not involved, version synchronicity is not necessary. If there are daemon protocol breakages, this is the fault of the Nix developers, not the fault of Nix users for trying out what is supposed to be a supported combination.

arianvp commented on pkgs/tools/package-management/nix-du/default.nix:
```diff
@@ -27,13 +27,13 @@ rustPlatform.buildRustPackage rec {
   doCheck = true;
   nativeCheckInputs = [
-    nix
+    nixForLinking
```
shouldn't this stay `nix`?  `nativeCheckInputs` aren't linked against.  We're calling the nix cli here; not linking against its library

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2025-02-24-nix-team-meeting-minutes-215-216/60920/1

SuperSandro2000 commented on pkgs/by-name/ni/nix-init/package.nix:
```diff
@@ -82,7 +82,7 @@ rustPlatform.buildRustPackage rec {
     # FIXME: our libgit2 is currently too new
     # LIBGIT2_NO_VENDOR = 1;
-    NIX = lib.getExe nix;
+    NIX = lib.getExe nixForLinking;
```
When looking at https://github.com/search?q=repo%3Anix-community%2Fnix-init+NIX+language%3ARust&type=code I don't think that is actually compiled in and if the env is not provided it just tried tries to execute "nix". Maybe we should move that to preCheck or so to make it clear.

9999years replied:
Ah, it is built in (`option_env!()` expands to the environment varible's value at compile time) but it's only using the `$NIX` variable for a command to execute, so we can keep this as `pkgs.nix`. Great catch, thanks!
https://doc.rust-lang.org/stable/std/macro.option_env.html

SuperSandro2000 commented on pkgs/by-name/nu/nurl/package.nix:
```diff
@@ -47,7 +47,7 @@ rustPlatform.buildRustPackage rec {
         lib.makeBinPath [
```
IMO we should suffix here. The user might have a git version in PATH with more support for things and wonders why nurl fails, eg ssh clones?

9999years replied:
I agree, but that's out of scope for this PR. Please open a PR with the suggested change and I'll approve it.

SuperSandro2000 commented:
> Wait a minute, it doesn't do this. In that chain of derivations, there is at least one fixed-output derivation. Overriding `nix` therefore doesn't cause a rebuild of chromium.

I think that is since #281902 where I had the same problem and needed to undo overrides and by splitting fixup-yarn-lock and prefetch-yarn-deps.

Ericson2314 commented:
> It sounds like we should just merge this as-is, since it benefits you and is very easy to replace later. Do you have any better suggestions for the name?

@lf- I am confused because I thought I *was* suggesting alternative which *also* allow making incremental progress, for which we have no objection:

- Whitelisting packages by making them use a new attribute for the metapackage works, and is an incremental process
- Making things that need a specific new version use `nix_x_y` works, and is an incremental process.

The intended workflow, as you say, is downstream you all *still* use the `nix = lix` overlay, to catch stragglers / 3rd party things, but over time time it is needed less and less.

The only issue I saw you raise with this was plugins... but then we agreed the *real* problem with plugins was various things enticing users from doing bad imperative plugin management for which they are likely to shoot themselves in the foot. So if we agree on that, then let's not making fixing the plugin situation a responsibility that the mechanism for swapping out Nix for Lix for be burdened with --- that's pure mission creep. Nix should fix up Nix's own messes, and do so in ways that Lix users need not think about at all.

With that resolved, I see no issues raised with the mechanisms I bulleted above.

-----

(BTW I realized the best precedent for thinking about plugins, `*ix-eval-jobs`, `hydra`, etc. is kernel modules, and `linuxPackages`. Those are DSOs if you squint, the kernel APIs are loudly and proudly unstable --- it's a perfect analogy. And the per-kernel package sets is also, of course, analogous to the mechanism we both agree should be used to deal with this stuff.)

Mic92 commented:
I am ok with nixForLinking actually. And I don't quite understand what @Ericson2314 are proposing? Do you want packages that link against a specific version to use nixVersions.nix_x_y?

9999years commented:
> @Mic92: Do you want packages that link against a specific version to use `nixVersions.nix_x_y`?

I believe this is what @Ericson2314 is suggesting, which seems to me like a bad idea for a couple reasons:

1. `nixVersions.nix_x_y` doesn't communicate intent. The intent is to link against `nixos/nix`, preferably the stable version.
2. `nixVersions.nix_x_y` obscures breakage and makes it harder to keep packages up to date. If package authors aren't notified when a Nix release breaks their build, their packages won't be fixed. As a result, it will become much harder to remove old `nixos/nix` versions from Nixpkgs. (This is one reason we still have Nix 2.3 in the tree.)
3. `nixVersions.nix_x_y` **makes it harder for the `nixos/nix` team to test new releases.** If there's not an obvious way to replace the `nixos/nix` version that packages are supposed to link against, there's not an obvious way to test releases to discover breakage.

Testability is one of the main motivations here — it's hard to test Lix HEAD if it breaks a bunch of packages that depend on the C++ API, and it's hard to test `nixos/nix` HEAD if it's difficult to upgrade (in a unified fashion via an overlay) packages that link against the `nixos/nix` C++ API.

9999years commented:
@Ericson2314 If you want to change the implementation here, you’re welcome to submit a PR. In the meantime, we plan to merge this PR as we believe it is an improvement on the status quo.

Mic92 commented:
> > @Mic92: Do you want packages that link against a specific version to use `nixVersions.nix_x_y`?
> 
> I believe this is what @Ericson2314 is suggesting, which seems to me like a bad idea for a couple reasons:

Agreed. I usually ping downstream maintainers on nix upgrades. I think pinning specific nix versions should the the exception rather than the norm (i.e. if a build cannot be fixed timely). We don't this for any other language (go, python, rust...)

Ericson2314 commented:
> So the libs from pkgs.nix have to have a different attr path than pkgs.nix in the future.

Yeah we could make that just the CLI and manual
